### PR TITLE
#166415129 feat database schema

### PR DESCRIPTION
**Pull request by despeauxz**
#### What does this PR do?
This PR designs the entire database schema for the application

#### Description of Task to be completed?
- create migrations
- create models
- associate each table

#### How should this be manually tested?
- run `npm run refreshdb`, should see a couple of tables being migrated in the database

#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[#166415129](https://www.pivotaltracker.com/story/show/166415129)
#### Screenshots (if appropriate)
N/A
#### Questions:
N/A

### CODE REVIEW

File: server/migrations/20190604155459-create-article.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    },
+    userId: {
+      type: Sequelize.INTEGER,
+      allowNull: false,
+      references: {
+        model: 'Users',
+        key: 'id'
+      }
+    },
+    slug: {
```
Comment (fidelisojeah) [inline server/migrations/20190604155459-create-article.js:17]: I guess this unique?

File: server/migrations/20190604161719-create-rating.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      type: Sequelize.INTEGER,
+      allowNull: false,
+      primaryKey: true,
+      references: {
+        model: 'Articles',
+        key: 'id',
+      },
+      onDelete: 'CASCADE',
+    },
+    ratings: {
```
Comment (fidelisojeah) [inline server/migrations/20190604161719-create-rating.js:27]: why do we allow null for this?

Also, how many times can a user rate an article?

if a user should be allowed to remove their rating from an article, consider using `paranoid:true` in the model definition (and having a `deletedAt` field here

File: server/models/article.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  Article.associate = (models) => {
+    const {
+      User, ArticleLike, Favourite, Rating, Comment, Tag, ReportArticle
+    } = models;
+    Article.belongsTo(User, {
+      foreignKey: 'userId',
+      as: 'author',
+    });
+
+    Article.hasMany(ArticleLike, {
```
Comment (fidelisojeah) [inline server/models/article.js:40]: some of your associations should actually be `belongToMany` through `userId`

File: server/models/permission.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
+module.exports = (sequelize, DataTypes) => {
+  const Permission = sequelize.define('Permission', {
+    access: {
+      type: DataTypes.STRING,
+      allowNull: false
+    }
+  }, {});
+
+  Permission.associate = (models) => {
```
Comment (fidelisojeah) [inline server/models/permission.js:13]: So users don't have permissions directly, users have roles and roles have permissions

File: server/models/passwordreset.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  const PasswordReset = sequelize.define('PasswordReset', {
+    userId: {
+      type: DataTypes.INTEGER,
+      allowNull: false
+    },
+    resetPasswordoken: {
+      type: DataTypes.STRING,
+      allowNull: true
+    },
+    resetPasswordExpiry: {
```
Comment (fidelisojeah) [inline server/models/passwordreset.js:11]: consider a `DateTime` field?

**Review by fidelisojeah — CHANGES_REQUESTED** [review]
This was always going to be really complicated.

please consider making some changes.

File: server/models/passwordreset.js (hunk 5 above)
Comment (despeauxz) [reply server/models/passwordreset.js:11]: I'll work on that, just got feedbacks

File: server/models/permission.js (hunk 4 above)
Comment (despeauxz) [reply server/models/permission.js:13]: should I create a joint table between the user and the permissions?

**Review by fidelisojeah — COMMENTED** [review]
pull and merge from develop.

**Review by EMacco — DISMISSED** [review]

**Review by fidelisojeah — DISMISSED** [review]

**Comment by GerrardE** [comment]
Good job, descriptive

**Review by EMacco — APPROVED** [review]

**Review by GerrardE — APPROVED** [review]

**Pull request merged**